### PR TITLE
Homework 2

### PR DESCRIPTION
**Pull request by sorokatayamv**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -99,5 +104,17 @@
...
+            <properties>
+                <testng.suite>${testng.configs.path}/exercisefirst.xml</testng.suite>
+            </properties>
+        </profile>
+        <profile>
+            <id>exerciseSecondTests</id>
+            <properties>
+                <testng.suite>${testng.configs.path}/exercisesecond.xml</testng.suite>
+            </properties>
+        </profile>
```
Comment (DmitryKhodakovsky) [inline pom.xml]: Why do you need two profiles instead of one for the hw2?

File: src/test/java/ru/training/at/hw2/TestBaseClass.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package ru.training.at.hw2;
+
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.chrome.ChromeDriver;
+import org.testng.annotations.AfterTest;
+import org.testng.annotations.BeforeTest;
+import org.testng.asserts.SoftAssert;
+
+public class TestBaseClass {
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw2/TestBaseClass.java]: should be `abstract`

File: src/test/java/ru/training/at/hw2/TestBaseClass.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    public WebDriver webDriver;
+
+    @BeforeTest
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw2/TestBaseClass.java]: Why do you use `@BeforeTest` and `@AfterTest`?

File: src/test/java/ru/training/at/hw2/TestBaseClass.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    public void beforeTest(){
+        System.setProperty("webdriver.chrome.driver", "C:\\chromedriver.exe");
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw2/TestBaseClass.java]: Why do you decide do not use `WebDriverManager` from boniagarcia for the managing of the driver?
Your test could be running only on Windows PC

File: src/test/java/ru/training/at/hw2/ex1/ExerciseFirst.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+public class ExerciseFirst extends TestBaseClass {
+
+    SoftAssert softAssert = new SoftAssert();
+
+    @BeforeTest
+    @Override
+    public void beforeTest() {
+        super.beforeTest();
+    }
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw2/ex1/ExerciseFirst.java]: What is the purpose of the overriding this method here?

File: src/test/java/ru/training/at/hw2/ex1/ExerciseFirst.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        loginButton.click();
+        WebElement username = (new WebDriverWait(webDriver, 10)
+                .until(ExpectedConditions.presenceOfElementLocated(By.id("name"))));
+        username.sendKeys("Roman");
+        WebElement password = (new WebDriverWait(webDriver, 10)
+                .until(ExpectedConditions.presenceOfElementLocated(By.id("password"))));
+        password.sendKeys("Jdi1234");
+        WebElement enterButton = (new WebDriverWait(webDriver, 10)
+                .until(ExpectedConditions.presenceOfElementLocated(By.xpath("//*[@id='login-button']"))));
+        enterButton.click();
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw2/ex1/ExerciseFirst.java]: May be it make sense create one variable for the `WebDriverWait` and do not create new instance of it every time?

File: src/test/java/ru/training/at/hw2/ex1/ExerciseFirst.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+                        .cssSelector("span[class='icons-benefit icon-custom']"))));
+        softAssert.assertTrue(imageSecond.isDisplayed());
+        WebElement imageThird = (new WebDriverWait(webDriver, 10)
+                .until(ExpectedConditions.presenceOfElementLocated(By
+                        .cssSelector("span[class='icons-benefit icon-multi']"))));
+        softAssert.assertTrue(imageThird.isDisplayed());
+        WebElement imageFourth = (new WebDriverWait(webDriver, 10)
+                .until(ExpectedConditions.presenceOfElementLocated(By
+                        .cssSelector("span[class='icons-benefit icon-base']"))));
+        softAssert.assertTrue(imageFourth.isDisplayed());
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw2/ex1/ExerciseFirst.java]: I suppose here elements could be collected to the `List<WebElement>` and you could use loop for the verification visibility of them

File: src/test/java/ru/training/at/hw2/ex1/ExerciseFirst.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+                        .xpath("//li[position()=3]"))));
+        softAssert.assertTrue(serviceLeft.isDisplayed());
+        WebElement metalsAndColorsLeft = (new WebDriverWait(webDriver, 10)
+                .until(ExpectedConditions.presenceOfElementLocated(By
+                        .cssSelector("a[href='metals-colors.html']"))));
+        softAssert.assertTrue(metalsAndColorsLeft.isDisplayed());
+        WebElement elementsPackLeft = (new WebDriverWait(webDriver, 10)
+                .until(ExpectedConditions.presenceOfElementLocated(By
+                        .xpath("//li[position()=5]"))));
+        softAssert.assertTrue(elementsPackLeft.isDisplayed());
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw2/ex1/ExerciseFirst.java]: `List<WebElement>`?

File: src/test/java/ru/training/at/hw2/ex1/ExerciseFirst.java [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    }
+    //12.Close Browser
+    @AfterTest
+    @Override
+    public void afterTest() {
+        super.afterTest();
+    }
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw2/ex1/ExerciseFirst.java]: What is the purpose of the overriding method here?

File: src/test/java/ru/training/at/hw2/ex2/ExerciseSecond.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+import org.testng.annotations.Test;
+import ru.training.at.hw2.TestBaseClass;
+
+public class ExerciseSecond extends TestBaseClass {
+
+    @BeforeTest
+    @Override
+    public void beforeTest() {
+        super.beforeTest();
+    }
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw2/ex2/ExerciseSecond.java]: What is the purpose of the overriding method here?

File: src/test/java/ru/training/at/hw2/ex1/ExerciseFirst.java [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+        softAssert.assertTrue(serviceLeft.isDisplayed());
+        WebElement metalsAndColorsLeft = (new WebDriverWait(webDriver, 10)
+                .until(ExpectedConditions.presenceOfElementLocated(By
+                        .cssSelector("a[href='metals-colors.html']"))));
+        softAssert.assertTrue(metalsAndColorsLeft.isDisplayed());
+        WebElement elementsPackLeft = (new WebDriverWait(webDriver, 10)
+                .until(ExpectedConditions.presenceOfElementLocated(By
+                        .xpath("//li[position()=5]"))));
+        softAssert.assertTrue(elementsPackLeft.isDisplayed());
+    }
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw2/ex1/ExerciseFirst.java:150]: You've forgot `softAssert.assertAll()`. You test will always passed

File: src/test/java/ru/training/at/hw2/ex2/ExerciseSecond.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+        loginButton.click();
+        WebElement username = (new WebDriverWait(webDriver, 10)
+                .until(ExpectedConditions.presenceOfElementLocated(By.id("name"))));
+        username.sendKeys("Roman");
+        WebElement password = (new WebDriverWait(webDriver, 10)
+                .until(ExpectedConditions.presenceOfElementLocated(By.id("password"))));
+        password.sendKeys("Jdi1234");
+        WebElement enterButton = (new WebDriverWait(webDriver, 10)
+                .until(ExpectedConditions.presenceOfElementLocated(By.xpath("//*[@id='login-button']"))));
+        enterButton.click();
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw2/ex2/ExerciseSecond.java]: May be it make sense create one variable for the WebDriverWait and do not create new instance of it every time?

File: src/test/java/ru/training/at/hw2/ex2/ExerciseSecond.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+                        .cssSelector("div.info-panel-body.info-panel-body-log > div > ul > li:nth-child(1)"))));
+        Assert.assertTrue(panelLogYellowDropdown.getText().contains("Colors: value changed to Yellow"));
+    }
+
+    //10.Close Browser
+    @AfterTest
+    @Override
+    public void afterTest() {
+        super.afterTest();
+    }
```
Comment (DmitryKhodakovsky) [inline src/test/java/ru/training/at/hw2/ex2/ExerciseSecond.java]: What is the purpose of the overriding method here?

**Review by DmitryKhodakovsky — CHANGES_REQUESTED** [review]
Please resolve comments

I am not able to run test on my PC
image

File: pom.xml (hunk 1 above)
Comment (sorokatayamv) [reply pom.xml]: I thought that for each class should be own profile

File: src/test/java/ru/training/at/hw2/ex1/ExerciseFirst.java (hunk 5 above)
Comment (sorokatayamv) [reply src/test/java/ru/training/at/hw2/ex1/ExerciseFirst.java]: I wanted to call parent's before method by this way. It was my night "right" thought))
But I have checked that parent's before and after methods would execute in any way. So overriding is unnecessary here.

File: src/test/java/ru/training/at/hw2/TestBaseClass.java (hunk 4 above)
Comment (sorokatayamv) [reply src/test/java/ru/training/at/hw2/TestBaseClass.java]: On the lecture was said that we can use this way too, so I decided use it. 
Resolved this comment and used WebDriverManager instead

**Review by DmitryKhodakovsky — APPROVED** [review]
HW2 accepted

**Comment by DmitryKhodakovsky** [comment]
Please resolve conflict and merge to the **master** branch HW2

**Pull request merged**